protocol/src/protocol.progress.proposed.md

```diff
@@ -0,0 +1,63 @@
+#### Reporting server task progress
+
+Many tools are capable of performing some background task processing or data streaming. From a UX point of view, it's good to report both the fact that the tool is performing some background work, but also report the progress being made for it. To realize that and to provide a simple proposal based on which the feature can be later improved, the following additions are proposed:
+
+_Client Capabilities_:
+
+The client sets the following capability if it is supporting notifying task progress.
+
+```ts
+/**
+ * Experimental client capabilities.
+ */
+experimental: {
+  /**
+   * The client has support for reporting progress.
+   */
+  progress?: boolean;
+}
+```
+
+##### Window Progress Notification
+_Notification_:
+
+The `window/progress` notification is sent from the server to the client to ask the client to indicate progress.
+
+* method: 'window/progress'
+* params: `ProgressParams` defined as follows:
+```ts
+export interface ProgressParams {
+    /**
+     * A unique identifier to associate multiple progress notifications with
+     * the same progress.
+     */
+    id: string;
+
+    /**
+     * Mandatory title of the progress operation. Used to briefly inform about
+     * the kind of operation being performed.
+     * Examples: "Indexing" or "Linking dependencies".
+     */
+    title: string;
+
+    /**
+     * Optional, more detailed associated progress message. Contains
+     * complementary information to the `title`.
+     * Examples: "3/25 files", "project/src/module2", "node_modules/some_dep".
+     * If unset, the previous progress message (if any) is still valid.
+     */
+    message?: string;
+
+    /**
+     * Optional progress percentage to display (value 100 is considered 100%).
+     * If unset, the previous progress percentage (if any) is still valid.
+     */
+    percentage?: number;
```

Comment:
I am not sure if percentage is a good idea since users could jump form 60 -> 50 which a lot of progress API's don't like. Could we used `worked`, `step` or `increment` instead.

Reply:
And should we defined that `-1` as a start value us infinite progress. I am also happy with an or type or an additional property.

Reply:
> I am not sure if percentage is a good idea since users could jump form 60 -> 50 which a lot of progress API's don't like. Could we used worked, step or increment instead.

My feeling is that just makes the API more complex. Usually the progress comes from some number of work done divided by some number of total work, e.g. bytes downloaded by value of `Content-Length` header. Having to calculate that, then diff it against the last reported progress percentage just because the API expects a delta makes it more complex on the server.

The client, if rendering a progress bar, ultimately needs to set the width of some element, which is easy using the latest absolute value. A delta API means the previous percentage value always needs to be kept around and incremented.

> which a lot of progress API's don't like

I don't know about other LSP clients, but which APIs are you thinking of?

Also some progress might actually sometimes be based on an estimate that may have to be corrected, decreasing the progress slightly.

> And should we defined that -1 as a start value us infinite progress. I am also happy with an or type or an additional property.

Could you explain what you mean by that? Isn't "Infinite progress" the same as "done", for which there is the `done: true` flag?

Reply:
> Could we used worked, step or increment instead.

I'd say if a client does not support "time-traveling" progress they can do a quick sanity check `this.progress = max(newProgress, thisProgress);`

> And should we defined that -1 as a start value us infinite progress.

Maybe it'd make sense to omit the `percentage` altogether in a server message when reporting an indefinite progress?